File: 2015/MAC0327/p40.cpp
#include<bits/stdc++.h>

using namespace std;

typedef struct Operacao {
    long long int l;
	long long int r;
	long long int d;
} operacao;

static const long long int MAXX = 1e5 + 1;

void le_operacao(operacao &op) {
    cin >> op.l >> op.r >> op.d;
	--(op.l);
	--(op.r);
}

int main() {
    operacao op[MAXX];
    long long int a[MAXX];
    long long int variacao[MAXX];
    long long int uso[MAXX];

	memset( op, 0, MAXX * sizeof (operacao));
	memset( a, 0, MAXX * sizeof (long long int));
	memset( variacao, 0, MAXX * sizeof (long long int));
	memset( uso, 0, MAXX * sizeof (long long int));
	long long int n, m, k;
	cin >> n >> m >> k;
	// Lendo o array inicial.
	for (long long int i = 0; i < n; i++) cin >> a[i];
	for (long long int i = 0; i < m; i++) le_operacao(op[i]);

	// Lendo os usos das operações
	for (long long int i = 0; i < k; i++) {
		long long int x, y;
		cin >> x >> y;
		uso[--x]++;
		if (--y != m - 1) uso[y + 1]--;
	}

	variacao[op[0].l] += uso[0] * op[0].d;
	if (op[0].r != n - 1) variacao[op[0].r + 1] -= uso[0] * op[0].d;
	for (long long int i = 1; i < m; i++) {
		uso[i] += uso[i - 1];
		variacao[op[i].l] += uso[i] * op[i].d;
		if (op[i].r != n - 1) variacao[op[i].r + 1] -= uso[i] * op[i].d;
	}

	cout << a[0] + variacao[0];
	for (long long int i = 1; i < n; i++) {
		variacao[i] += variacao[i - 1];
		cout << " " << a[i] + variacao[i];
	}
	cout << endl;
	return 0;
}

File: 2015/MAC0327/p40.py
def main():
    n, m, k = map(int, raw_input().split())
    array = map(int, raw_input().split())
    operacoes = []

    for i in range(m):
        operacoes.append(map(int, raw_input().split()))



if __name__ == '__main__':
    main()



#include<bits/stdc++.h>

using namespace std;

/******************************************************************************/

class operation {
public:
	long long int l;
	long long int r;
	long long int d;
};

int main() {
	memset( op, 0, MAXX * sizeof (operation));
	memset( a, 0, MAXX * sizeof (long long int));
	memset( difa, 0, MAXX * sizeof (long long int));
	memset( difo, 0, MAXX * sizeof (long long int));
	long long int n, m, k;
	cin >> n >> m >> k;
	cin >> a[0];
	for (long long int i = 1; i < n; i++) {
		cin >> a[i];
	}
	for (long long int i = 0; i < m; i++) {
		long long int l, r;
		cin >> l >> r >> op[i].d;
		op[i].l = --l;
		op[i].r = --r;
	}
	for (long long int i = 0; i < k; i++) {
		long long int x, y;
		cin >> x >> y;
		--x;
		--y;
		difo[x]++;
		if (y != m - 1) {
			difo[y + 1]--;
		}
	} // aqui difo[] e' o vet diferenca da freq de uso das operacoes op[]

	difa[op[0].l] += difo[0] * op[0].d;
	if (op[0].r != n - 1) {
		difa[op[0].r + 1] -= difo[0] * op[0].d;
	}
	for (long long int i = 1; i < m; i++) {
		difo[i] += difo[i - 1]; // iterativamente recupera o vetor op atualizado (onde difo[0] == op[0] atualizado)
		difa[op[i].l] += difo[i] * op[i].d;
		if (op[i].r != n - 1) {
			difa[op[i].r + 1] -= difo[i] * op[i].d;
		}
	}

	cout << a[0] + difa[0];
	for (long long int i = 1; i < n; i++) {
		difa[i] += difa[i - 1];
		cout << " " << a[i] + difa[i];
	}
	cout << endl;
	return 0;
}